docs/Migration Guide.md

```diff
@@ -3,6 +3,154 @@
+migration:
+
+```c++
+template <typename T>
+inline T* MyGetMaybePtrFromReservedSlot(JSObject* obj, size_t slot) {
+    JS::Value v = JS::GetReservedSlot(obj, slot);
+    return v.isUndefined() ? nullptr : static_cast<T*>(v.toPrivate());
+}
+```
+
+Additionally, if you have any `JS::Heap<>` pointers to GC things as
+members of your private C++ structs, this migration might be a good time
+to consider moving those GC things into their own reserved slots on the
+object, and just keep C++ data in the C++ struct.
+This technique might not be right for every codebase, but it's worth
+considering.
+
+### Property keys ###
+
+As in the previous migration from 78 to 91, more JSID macros have now
+been replaced by methods of `JS::PropertyKey`.
+
+**Recommendation:** Before the migration, replace `JSID_IS_VOID` (or an
+equality comparison to `JSID_VOID`) with `id.isVoid()`.
+
+After the migration, replace the following:
+- `JSID_VOID` → `JS::PropertyKey::Void()`
+- `SYMBOL_TO_JSID()` → `JS::PropertyKey::Symbol()`
+- `JSID_BITS(id)` → `id.asRawBits()`
+- `JSID_TO_LINEAR_STRING(id)` → `id.toLinearString()`
+
+### Headers ###
+
+There are now more headers which should be included separately in code
+which uses their functionality.
+
+This is a list of common ones that might be used by embeddings, but in
+general if you are missing function definitions when compiling your
+code, try checking if you might have to include another header.
+
+- `JS::Call`, `JS::Construct`, and related — `<js/CallAndConstruct.h>`
+- `JS_DefineDebuggerObject()` — `<js/Debug.h>`
+- `JS::CurrentGlobalOrNull()` — `<js/GlobalObject.h>`
+- `JS_DefineProperty()`, `JS_DefineFunction()`, `JS_Enumerate()`,
+  `JS_GetElement()`, `JS_GetProperty()`, `JS_HasProperty()`,
+  `JS_SetProperty()`, and many other functions related to accessing
+  properties or array elements — `<js/PropertyAndElement.h>`
+- `JS::GetScriptPrivate()` and `JS::SetScriptPrivate()` —
+  `<js/ScriptPrivate.h>`
+- `JS::BuildStackString()` and other functions related to the call stack
+  — `<js/Stack.h>`
+
+### Various API changes ###
+
+This is a non-exhaustive list of minor API changes and renames.
+
+- `JS::PropertyKey::fromNonIntAtom()` → `JS::PropertyKey::NonIntAtom()`
+- `JS::SafelyInitialized<T>()` → `JS::SafelyInitialized<T>::create()`
+- `JSFreeOp` → `JS::GCContext` (in particular, in finalize operations)
+- `JS_UpdateWeakPointerAfterGC()` and weak-pointer update callbacks now
+  take an additional `JSTracer*` argument.
+
+The following things have been removed.
+In several cases, you can make the necessary changes before the
+migration already.
+
+- `JS_AtomizeAndPinJSString()` has been removed, there is no direct
+  replacement. There are several `JS_AtomizeAndPin...` APIs that might
+  work instead.
+- `JS::CloneAndExecuteScript()` has been removed, there is no direct
+  replacement. Usually, `JS::Evaluate()`, or `JS::Compile()` followed by
+  `JS_ExecuteScript()`, should be sufficient.
```

Comment:
The clone functionality is removed in favor of `JS::Stencil` returned by `JS::CompileGlobalScriptToStencil`, and multiple `JS::InstantiateGlobalStencil` calls on `JS::Stencil` to get `JSScript` in each compartment/realm, and `JS_ExecuteScript` for each.

If the compilation cost isn't problematic, the options mentioned in the patch works.
If the compilation cost is problematic and that's the reason the embedding was using the clone functionality, the replacement is the `JS::Stencil` way.

I'll look into adding examples for `JS::Stencil` APIs.

Reply:
I saw #63 — thanks!